src/completion.ts

```diff
@@ -493,49 +493,75 @@ class MdCompletionItemProvider implements CompletionItemProvider {
+                                    RXlink + RXlinktitle +
+                                    String.raw`)`;  // end regex
+                const RXflags = String.raw`mg`;     // multiline & global
+                // This pattern matches linklabels in link references definitions:  [linklabel]: link "link title"
+                const pattern = new RegExp(RXlookbehind + RXlinklabel + RXlookahead, RXflags);
+                const refLabels = document.getText().match(pattern)!
+                    // Remove leading and trailing whitespace characters in each label.
+                    .map(str => str.replace(/^[ \t\r\n\f\v]+/, '').replace(/[ \t\r\n\f\v]+$/, ''));
+
+                // Remove duplicates. Perform case-insensitive comparison, but preserve original case in result.
+                // TODO: may be extracted to a seperate function and used for all completions in future
+                refLabels.sort();
+                const tidyRefLabels = refLabels.reduce<string[]>((result, crt) => {
+                    crt = crt.trim();
+                    if (
+                        (result.length > 0 && crt.toUpperCase() !== result[result.length - 1].toUpperCase())
+                        || result.length === 0
+                    ) {
+                        result.push(crt);
                     }
-                    return prev;
+                    return result;
                 }, []);
 
-                res(refLabels);
+                const completionItemList = tidyRefLabels.map<CompletionItem>(function (ref) {
+                    const item = new CompletionItem(ref, CompletionItemKind.Reference);
+                    const usages = usageCounts.get(ref) || 0;
+                    item.documentation = new MarkdownString(ref);
+                    item.detail = usages === 1 ? `1 usage` : `${usages} usages`;
+                    // Prefer unused items. <https://github.com/yzhang-gh/vscode-markdown/pull/414#discussion_r272807189>
+                    item.sortText = usages === 0 ? `0-${ref}` : `1-${ref}`;
+                    item.range = range;
+                    return item;
+                });
+                res(completionItemList);
             });
-        } else if (/\[[^\]]*\]\(#[^\)]*$/.test(lineTextBefore)) {
+        } else if (
+            /\[[^\]]*?\]\(#[^\)]*$/.test(lineTextBefore)
+            // /\[[^\]]*\]\((\S*)#[^\)]*$/.test(lineTextBefore) // `[](url#anchor|` Link with anchor.
+            // || /\[[^\]]*\]\:\s?(\S*)#$/.test(lineTextBefore) // `[]: url#anchor|` Link reference definition with anchor.
+        ) {
             /* ┌───────────────────────────┐
                │ Anchor tags from headings │
                └───────────────────────────┘ */
-            let startIndex = lineTextBefore.lastIndexOf('(');
+            let startIndex = lineTextBefore.lastIndexOf('#') - 1;
+            let isLinkRefDefinition = /^[>]? {0,3}\[[^\]]+?\]\:[ \t\f\v]*#*$/.test(lineTextBefore);
```

Comment:
Looking at it here I'm not sure why its #* and not just #. Maybe I was thinking n-level headers would have multiple #'s but that's incorrect, it would only ever be singular.

Reply:
Its history:

1. `/\[[^\]]*\]\:\s?(\S*)#*$/` (8369b7a8bb3769cbe5b79b2e21f61e13fd1ae28b)
2. `/^[>]? {0,3}\[[^\]]+?\]\:[ \t\f\v]*#*$/` (12a500ab96e31b286c0b81a287457bd6f2431603)

Additionally, b4807d0bc221a1611a56a09439dbd26a714aacbe changed a regexp near it.

Seems it could be `#[^#]*$`.

Reply:
That would seem to make sense. I'll not be able to have a proper look till this evening (UK time)